### ISSUE TITLE
Add a suitable {-# WARNING in "x-partial" ... #-} to Data.List.{init,last}

### ISSUE THREAD

**Issue by mpilgrem** [haskell/core-libraries-committee#292]
Following the past decisions in respect of `Data.List.head` and `Data.List.tail` in:
* #87 
* #114

implemented in `base-4.19.0.0` (GHC 9.8.1), which also added `Data.List.unsnoc`, I propose, for consistency, that the same idea is extended to `Data.List.init` and `Data.List.last`. Their Haddock documentation has warned that they are partial and suggested alternatives since `base-4.17.0.0` (GHC 9.4.1).

The basis for the proposal is purely consistency of treatment of `Data.List.{head,tail,init,last}`. It seems to me that the principal merits and drawbacks applicable to `head` and `tail`, discussed at some length in the past approved proposals, are equally applicable to `init` and `last`. 

@Bodigrim, on introducing #87 (at https://github.com/haskell/core-libraries-committee/issues/87#issue-1369083584), gave as a reason that it was not extended then to `init` or `last` that doing so would require `Data.List.unsnoc` first. That barrier no longer applies.

In case it is relevant, I note:
1.  that other (total) functions in `base` are defined in terms of `Data.List.init` or `Data.List.last`. For example, `Data.List.NonEmpty.init`:
    ~~~haskell
    init :: NonEmpty a -> [a]
    init ~(a :| as) = List.init (a : as)
    ~~~
2.  that Cabal's autogenerated `Paths_<package_name>.hs` module makes use of `Data.List.last` - see https://hackage.haskell.org/package/Cabal-3.14.0.0/src/src/Distribution/Simple/Build/PathsModule/Z.hs.

EDIT: The resolution of this issue may depend on the resolution of the following issue, given that this issue provides advice (as does the existing Haddock documentation) about the use of `unsnoc`:
* #307

**Comment by phadej** [comment]
The `Paths_*.hs` usage originates from `filepath` https://github.com/haskell/filepath/blob/e794e7376755fdaecae6fc23bd34556d46c46307/System/FilePath/Internal.hs#L747 (joinFilePath is "inlined" to have only dependencies on `base`).

It feels that 

```haskell
foo [] = something
foo xs = somethingElseWith (last xs)
```

it not uncommon pattern in older code, because writing that safely was quite inconvenient using just `base`. Now we can rewrite that as

```haskell
foo [] = something
foo (x : xs) = somethingElseWith (NonEmpty.last (x :| xs))
```

i.e. above `Data.List.init` could be written in terms of `Data.List.NonEmpty.init`, not other way around; That refactoring may cause some module reorganization to avoid circular dependencies.

---

It would be really helpful to see in how many modules in `base` one would actually need to disable `x-partial` warning, if it's added to `init` and `last`. In other words, make a MR.

**Comment by Bodigrim** [comment]
Nowadays one can write `foo = maybe something (somethingElseWith . snd) . Data.List.unsnoc`.

**Comment by phadej** [comment]
`unsnoc` is since base-4.19, i.e. GHC-9.8; quite recent addition. On the other hand from https://github.com/haskell/core-libraries-committee/issues/228#issuecomment-2170037180

> In April 2024 CLC approved https://github.com/haskell/core-libraries-committee/issues/262 to clarify that we do not recognise warnings as breaking changes, dilluting meaning of 3-release policy even further.

---

In that light, simply adding a proposed warning seems to be inline with recent-ish CLC views. `filepath` and `Cabal` and others just need to adapt, or disable warnings. No big deal.

**Comment by Bodigrim** [comment]
We are past GHC 9.12 / base-4.21 fork, so the proposed warning can land no sooner than base-4.22, making `unsnoc` compatible with the 3-release policy, even in its stricter version.

**Comment by Bodigrim** [comment]
@mpilgrem would you like to prepare an MR for this? 
(Besides boot libraries, you'd have to clean up `init` and `last` throughout GHC itself or, alternatively, tweak `GHC.Prelude.Basic` to re-export them without warnings)

**Comment by mpilgrem** [comment]
@Bodigrim, I have created:
* https://gitlab.haskell.org/ghc/ghc/-/merge_requests/13564

On avoiding warnings within the GHC repository, I added `{-# OPTIONS_GHC -Wno-x-partial #-}` in various places, following the existing example for `head` used in module `GHC.Internal.Control.Monad.Fix`. I think the task of the respository changing code to heed the warnings in `base` is a distinct task.

**Comment by mpilgrem** [comment]
@Bodigrim, may I ask for a bit a advice? I have run into package `time` using `last`, so the CI pipeline of my `ghc/ghc` respository MR will not pass.

There is a `time` repository at https://github.com/haskell/time (referenced on Hackage). However, it seems to me that the https://gitlab.haskell.org/ghc/ghc/ repository may refer to https://gitlab.haskell.org/ghc/packages/time at a specific commit (because there is a `time @ e5c5d198` file under https://gitlab.haskell.org/ghc/ghc/libraries that goes there when I click it).

I am not sure _what_ I should be proposing to change, to try to make the changes that will allow GHC's CI to pass for my MR. I think that all is needed, in the immediate instance, is for the relevant module in `time` to have the addition of:
~~~haskell
{-# LANGUAGE CPP #-}

#if MIN_VERSION_GLASGOW_HASKELL(9,8,1,0)
-- For init and last in trimTrailing
{-# OPTIONS_GHC -Wno-x-partial #-}
#endif
~~~

**Comment by Bodigrim** [comment]
1. You don't need CPP, the incantation is `{-# OPTIONS_GHC -Wno-unrecognized-warning-flags -Wno-x-partial #-}`. 

2.  https://gitlab.haskell.org/ghc/packages/time  is a mirror of https://github.com/haskell/time, it will be autoupdated once the source repo is.

3. Instead of updating `time`, you can pass `-Wno-x-partial` via `hadrian/src/Settings/Warnings.hs`.

**Comment by meooow25** [comment]
I see that the proposed warning directs users to `unsnoc`, but there is a performance penalty to using `uncons` in place of `init` or `last`, because of all the `Maybe`s and `(,)`s. It is particularly bad for `last` because allocations increase from zero to O(n).

If you would like to check the Core: https://play.haskell.org/saved/ySiaRuho

**Comment by mpilgrem** [comment]
For information: the origin of the text of the proposed compiler warning is the existing warning in the Haddock documentation for `init` and `last` in `base >= 4.19.0.0` (which is unqualified as to performance).

**Comment by mpilgrem** [comment]
I have no experience in measuring, or optimising, Haskell code performance, but (in the light of @meooow25's comments) I tried to apply the `criterion` library in the following example:
~~~haskell
module Main (main) where

import Criterion.Main
import Data.List ( unsnoc )
import Data.List.NonEmpty ( NonEmpty ( ..) )
import qualified Data.List.NonEmpty as NE

partial :: String -> Maybe (String, Char)
partial "" = Nothing
partial s = Just (init s, last s)

total1 :: String -> Maybe (String, Char)
total1 = unsnoc

total2 :: String -> Maybe (String, Char)
total2 "" = Nothing
total2 (c:cs) = let s = c :| cs in Just (NE.init s, NE.last s)

main :: IO ()
main = defaultMain
  [ bgroup "last" [ bench "partial" $ nf partial "Hello! World"
                  , bench "total1" $ nf total1 "Hello! World"
                  , bench "total2" $ nf total2 "Hello! World"
                  ]
  ]

~~~

The results were typically like the following:
~~~text
benchmarking last/partial
time                 164.2 ns   (161.4 ns .. 169.2 ns)
                     0.992 R²   (0.988 R² .. 0.996 R²)
mean                 179.7 ns   (174.9 ns .. 185.7 ns)
std dev              18.33 ns   (16.79 ns .. 20.59 ns)
variance introduced by outliers: 91% (severely inflated)

benchmarking last/total1
time                 392.6 ns   (377.3 ns .. 405.8 ns)
                     0.993 R²   (0.989 R² .. 0.996 R²)
mean                 368.4 ns   (361.7 ns .. 377.5 ns)
std dev              26.58 ns   (19.20 ns .. 36.09 ns)
variance introduced by outliers: 82% (severely inflated)

benchmarking last/total2
time                 167.8 ns   (164.8 ns .. 171.7 ns)
                     0.996 R²   (0.994 R² .. 0.999 R²)
mean                 169.3 ns   (166.8 ns .. 174.0 ns)
std dev              11.74 ns   (7.172 ns .. 20.49 ns)
variance introduced by outliers: 82% (severely inflated)
~~~

I was surprised that `unsnoc` ('total1') was so much worse than 'total2'. It made me wonder why 'unsnoc' is not specified like 'total2'. 'unsnoc' has a code comment: `-- Expressing the recursion via 'foldr' provides for list fusion.` Perhaps that has something to do with it or perhaps my example does not reflect how `init` and `last` are used 'in the wild'.

**Comment by mpilgrem** [comment]
I repeated my `criterion` experiment with polymorphic functions and, instead of `"Hello! World"`, a `list :: [Int]; list = replicate 10000 1`. The results had the same pattern:
~~~test
benchmarking last/partial
time                 78.17 μs   (75.80 μs .. 80.59 μs)
                     0.994 R²   (0.992 R² .. 0.997 R²)
mean                 77.59 μs   (76.08 μs .. 79.32 μs)
std dev              5.037 μs   (4.412 μs .. 5.981 μs)
variance introduced by outliers: 66% (severely inflated)

benchmarking last/total1
time                 332.6 μs   (318.8 μs .. 349.7 μs)
                     0.991 R²   (0.986 R² .. 0.997 R²)
mean                 317.6 μs   (313.6 μs .. 325.3 μs)
std dev              18.25 μs   (12.06 μs .. 28.19 μs)
variance introduced by outliers: 53% (severely inflated)

benchmarking last/total2
time                 79.66 μs   (77.45 μs .. 81.44 μs)
                     0.996 R²   (0.994 R² .. 0.998 R²)
mean                 79.62 μs   (78.34 μs .. 81.08 μs)
std dev              4.555 μs   (3.885 μs .. 5.611 μs)
variance introduced by outliers: 60% (severely inflated)
~~~

**Comment by mpilgrem** [comment]
By way of update, the merge request pipeline for the MR is now all green (following some false starts).

The existing `unsnoc` performance issue remains. @Bodigrim, as the originator of #165, may I draw your attention to the discussion above? In short, if the tests above are not wrong, it seems existing `unsnoc` is somewhat slower than the 'naïve':
~~~haskell
import Data.List.NonEmpty ( NonEmpty (..) )
import qualified Data.List.NonEmpty as NE

unsnoc :: [a] -> Maybe ([a], a)
unsnoc [] = Nothing
unsnoc (x:xs) = let l = x :| xs in Just (NE.init l, NE.last l)
~~~

If they are not wrong, it may be that there should be a new issue for the CLC on that topic, with this issue dependent on the resolution of that issue.

**Comment by Bodigrim** [comment]
> I see that the proposed warning directs users to `unsnoc`, but there is a performance penalty to using `unsnoc` in place of `init` or `last`, because of all the `Maybe`s and `(,)`s. It is particularly bad for `last` because allocations increase from zero to O(n).

These are cheap, short-living allocations in the nursery, so they are not materially worse than O(n) time.

----

@mpilgrem The existing implementation of `unsnoc` wraps and unwraps both `Maybe` and `(,)` on each iteration, which is expensive and probably explains your benchmarks. If we make a special case for `unsnoc [] = Nothing` (thus sacrificing list fusion), we can avoid wrapping / unwrapping of `Maybe`. If we additionally sacrifice the requirement of traversing the list only once, we can indeed avoid wrapping / unwrapping of `(,)` as in `Just (NE.init l, NE.last l)` above. 

Sacrificing list fusion is probably reasonable: in cases where list fusion used to happen we will lose one allocation on materializing `(:)` but save one allocation on `Maybe`, which is no worse; and we simply win in cases where fusion didn't use to happen.

Traversing the list twice is less palatable to my taste, because it can double maximum memory residence: we would materialise the input in full while computing `NE.init l`, then keep both `l` and `NE.init l` in memory while computing `NE.last l` (after which `l` can be finally freed).

Could you possibly benchmark an implementation of `unsnoc` with `unsnoc [] = Nothing; unsnoc xs = Just $ go xs`? Is it far off the mark?

**Comment by mpilgrem** [comment]
@Bodigrim, I did not follow what `go` function you intended, but I tried this:
~~~haskell
unsnocGo :: [a] -> Maybe ([a], a)
unsnocGo [] = Nothing
unsnocGo (x : xs) = Just $ go (x :| xs)
 where
  go (y :| []) = ([], y)
  go (y1 :| (y2 : ys)) = (\((a, b)) -> (y1 : a, b)) (go (y2 :| ys))
{-# INLINABLE unsnocGo #-}
~~~

I also noticed, by accident, for the `base` `unsnoc` if `~(a, b)` is replaced by just `(a, b)` it is about 15% faster.

My complete experiment (with GHC 9.8.3) is below. I am now using a long list (`[1 .. 10000] :: Int`). I copied the `unsnoc` from `base`, rather than importing it.
~~~haskell
module Main (main) where

import Criterion.Main
import Data.List.NonEmpty ( NonEmpty (..) )
import qualified Data.List.NonEmpty as NE

partial :: [a] -> Maybe ([a], a)
partial [] = Nothing
partial s = Just (init s, last s)

naive ::  [a] -> Maybe ([a], a)
naive [] = Nothing
naive (x : xs) = let l = x :| xs in Just ( NE.init l, NE.last l)

unsnoc :: [a] -> Maybe ([a], a)
unsnoc = foldr (\x -> Just . maybe ([], x) (\(~(a, b)) -> (x : a, b))) Nothing
{-# INLINABLE unsnoc #-}

unsnocNoTilde :: [a] -> Maybe ([a], a)
unsnocNoTilde = foldr (\x -> Just . maybe ([], x) (\((a, b)) -> (x : a, b))) Nothing
{-# INLINABLE unsnocNoTilde #-}

unsnocGo :: [a] -> Maybe ([a], a)
unsnocGo [] = Nothing
unsnocGo (x : xs) = Just $ go (x :| xs)
 where
  go (y :| []) = ([], y)
  go (y1 :| (y2 : ys)) = (\((a, b)) -> (y1 : a, b)) (go (y2 :| ys))
{-# INLINABLE unsnocGo #-}

list :: [Int]
list = [1 .. 10000]

main :: IO ()
main = defaultMain
  [ bgroup "unsnoc" [ bench "partial" $ nf partial list
                    , bench "naive" $ nf naive list
                    , bench "unsnoc" $ nf unsnoc list
                    , bench "unsnocNoTilde" $ nf unsnocNoTilde list
                    , bench "unsnocGo" $ nf unsnocGo list
                    ]
  ]
~~~

Typical results are:
~~~text
benchmarking unsnoc/partial
time                 86.54 μs   (83.61 μs .. 89.08 μs)
                     0.992 R²   (0.989 R² .. 0.995 R²)
mean                 87.59 μs   (85.35 μs .. 91.38 μs)
std dev              10.66 μs   (5.786 μs .. 17.82 μs)
variance introduced by outliers: 88% (severely inflated)

benchmarking unsnoc/naive
time                 83.51 μs   (81.15 μs .. 86.66 μs)
                     0.994 R²   (0.993 R² .. 0.996 R²)
mean                 85.34 μs   (83.90 μs .. 86.86 μs)
std dev              4.928 μs   (4.369 μs .. 5.827 μs)
variance introduced by outliers: 60% (severely inflated)

benchmarking unsnoc/unsnoc
time                 320.9 μs   (315.8 μs .. 326.5 μs)
                     0.998 R²   (0.997 R² .. 0.999 R²)
mean                 320.9 μs   (317.6 μs .. 325.3 μs)
std dev              12.70 μs   (8.529 μs .. 16.35 μs)
variance introduced by outliers: 35% (moderately inflated)

benchmarking unsnoc/unsnocNoTilde
time                 269.5 μs   (265.0 μs .. 278.3 μs)
                     0.996 R²   (0.991 R² .. 0.999 R²)
mean                 266.3 μs   (263.9 μs .. 270.5 μs)
std dev              11.94 μs   (8.054 μs .. 20.08 μs)
variance introduced by outliers: 42% (moderately inflated)

benchmarking unsnoc/unsnocGo
time                 115.9 μs   (114.1 μs .. 117.9 μs)
                     0.998 R²   (0.997 R² .. 0.999 R²)
mean                 114.8 μs   (113.9 μs .. 116.4 μs)
std dev              3.864 μs   (2.812 μs .. 5.289 μs)
variance introduced by outliers: 32% (moderately inflated)
~~~

That is, the 'naive' total version is about as quick as a 'partial' version (as before). The copy of `base` `unsnoc` is well over three times slower (as before). Dropping the tilde saves about 15% in time. Finally, my 'go' version is about 1.4 times slower than the 'naive' total version.

**Comment by mpilgrem** [comment]
@meooow25, @Bodigrim I have separated out the question of the time performance of existing `Data.List.unsnoc` into its own issue at:
* #307

**Comment by Bodigrim** [comment]
@mpilgrem given that #307 was withdrawn, how would you like to proceed here?

**Comment by mpilgrem** [comment]
@Bodigrim, sorry for the hiatus. I had to focus on work.

To recap:

* existing (`base-4.21.0.0`) points users of partial `head` and `tail` to `uncons` (among other things);
* my [MR](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/13564) added similar advice for partial `init` and `last`;
* @meooow25 suggested that `uncons` may be poor advice in many circumstances, because (lazy) `unsnoc` has poor time performance. @meooow25 provided a [pecking-order](https://github.com/haskell/core-libraries-committee/issues/307#issuecomment-2548958498) of alternatives.

If the CLC can tolerate this issue #292 being open for a little while longer, I am going to make a separate proposal to introduce time-performant alternatives to `unsnoc` which, if accepted, would allow partial `head`, `tail`, `init` and `last` give more tailored advice to users.

**Comment by Bodigrim** [comment]
IMHO the performance of `unsnoc` is not a pressing concern. If a constant-factor penalty from replacing `init` / `last` by `unsnoc` is a bottleneck for your application, surely you'd rather worry about using a linear-time operation where constant time is possible with a more suitable data type than lists. (But I'm no longer a member of CLC)

**Comment by hasufell** [comment]
I see those things as separate concerns:

* warning about unsafe functions is always good (even if the alternatives are less performant... that's actually quite often the case)
* coming up with more performant alternatives to `unsnoc`

I'd suggest to:

- add the warning to `init` and `last`
- improve documentation of `unsnoc` to point to its performance intricacies (this does not need a CLC proposal)
- write more performant alternatives for the above (a separate proposal for the future)

**Comment by mpilgrem** [comment]
As a recap of the current position: currently (`base-4.21.0.0`), `head` and `tail` warn (in the GHC sense) and `init` and `last` advise (in Haddock documentation) (respectively):
* `head`: "This is a partial function, it throws an error on empty lists. Use pattern matching, `uncons` or `listToMaybe` instead. Consider refactoring to use `Data.List.NonEmpty`."
* `tail`: "This is a partial function, it throws an error on empty lists. Replace it with `drop 1`, or use pattern matching or `uncons` instead. Consider refactoring to use `Data.List.NonEmpty`."
* `init`: "WARNING: This function is partial. Consider using `unsnoc` instead."
* `last`: "WARNING: This function is partial. Consider using `unsnoc` instead."

and my GHC merge request [#13564](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/13564) proposes the following GHC warnings for `init` and `last`:
* `init`: "This is a partial function, it throws an error on empty lists. Use `unsnoc` instead. Consider refactoring to use `Data.List.NonEmpty`."
* `last`: "This is a partial function, it throws an error on empty lists. Use `unsnoc` instead. Consider refactoring to use `Data.List.NonEmpty`."

**Comment by hasufell** [comment]
@mpilgrem so you agree to keep the scope of this proposal to just the warnings and move on to a vote?

**Comment by hasufell** [comment]
Are there any non-binding opinions or alternative suggestions as to the exact nature of the warning, before we move to a proper vote?

The warnings mention `unsnoc` and `Data.List.NonEmpty` as alternatives:

```hs
{-# WARNING in "x-partial" last "This is a partial function, it throws an error on empty lists. Use 'GHC.Internal.Data.List.unsnoc' instead. Consider refactoring to use \"Data.List.NonEmpty\"." #-}

{-# WARNING in "x-partial" init "This is a partial function, it throws an error on empty lists. Use 'Data.List.unsnoc' instead. Consider refactoring to use \"Data.List.NonEmpty\"." #-}
```

----

@angerman @parsonsmatt @velveteer @TeofilC @Daniel-Diaz @ChickenProp @noughtmare @mpscholten @doyougnu

**Comment by TeofilC** [comment]
You shouldn't refer to `GHC.Internal.Data.List`. The reference should go to the public facing identifier instead. If that's not possible then it sounds like a Haddock bug.

Otherwise this sounds like a good change to me.

**Comment by mpilgrem** [comment]
@TeofilC, sorry - that (`GHC.Internal.Data.List`) is my mistake, I think - I'll fix.

**Comment by mpscholten** [comment]
👍 looks good to me

**Comment by Daniel-Diaz** [comment]
I agree with @ChickenProp that "consider using" is preferable over just "use", as `unsnoc` doesn't seem to be a better alternative in every way or in every circumstance.

I also think that the compiler can and should suggest, but that, ultimately, the decision falls into the programmer. The language should reflect that.

That being said, two "consider" in a row sounds a bit repetitive. Perhaps:

```
This is a partial function. It throws an error on empty lists. Consider using 'Data.List.unsnoc' instead or refactoring to use \"Data.List.NonEmpty\"."
```

Or even more concise:

```
This function throws an error on empty lists. Consider using 'Data.List.unsnoc' instead or refactoring to use \"Data.List.NonEmpty\"."
```

**Comment by angerman** [comment]
Adding warnings seems to be a good idea to me. +1

**Comment by hasufell** [comment]
@mpilgrem do you wish to take the advice and adjust the warning message? We can move to a proper vote quickly then.

**Comment by mpilgrem** [comment]
@hasufell, if the CLC come to a settled view on the appropriate approach to the messages, then I'll, of course, reflect that in the associated merge request.

By way of explanation of why it is currently put as it is, the motivating issue was consistency of approach with `head`, `tail`. They are _directive_ in their messages about non-partial alternatives (provided by Haskell or the `base` package) for the _same_ type and _suggestive_ in their messages about refactoring to use a _different_ type. For example:

`head`: "This is a partial function, it throws an error on empty lists. Use [DIRECTIVE] pattern matching, `uncons` or `listToMaybe` instead. Consider [SUGGESTIVE] refactoring to use `Data.List.NonEmpty`."

It may be more stark with `init`, `last` because there is only one non-partial alternative for the same type, currently (this may change, sooner rather than later, see #316).

If the CLC settles on a suggestive tone for all alternatives for `init`, `last`, does this require a change in the existing approach taken to `head`, `tail`? If the answer is 'yes', then I'm happy to put that through the same merge request. The example of `head` may become:

`head`: "This is a partial function, it throws an error on empty lists. Consider [SUGGESTIVE] using pattern matching, `uncons` or `listToMaybe` instead; or refactoring to use `Data.List.NonEmpty`."

**Comment by Bodigrim** [comment]
The original reason for "use" vs. "consider" was that "use" listed immediate replacements, while "consider" was for non-local refactoring. E. g., you can use `drop 1` instead of `tail` - this is a local refactoring, immediately available to silence the warning, although admittedly somewhat obscuring the matter. Ideally you should strive to express invariants in types and use `NonEmpty` - but it might incur lots of changes in other parts of your program, so maybe you don't want to do this right now, thus "consider".

While my personal preference remains to be "Use ..., consider ..." pattern, I would not dwell on this much. I doubt any reader to be alert enough to such subtlety.

**Comment by treeowl** [comment]
Could we get a `Data.List Partial` that exports versions of all the partial functions without warnings?

**Comment by hasufell** [comment]
> if the CLC come to a settled view on the appropriate approach to the messages, then I'll, of course, reflect that in the associated merge request

I personally don't really care for the exact wording and it seemed to me the others were merely suggestions that you are free to adjust to or leave it as is.

But let us double check: @Daniel-Diaz do you feel strongly about the wording? Otherwise I want to move to a vote.

**Comment by Daniel-Diaz** [comment]
> > if the CLC come to a settled view on the appropriate approach to the messages, then I'll, of course, reflect that in the associated merge request
> 
> I personally don't really care for the exact wording and it seemed to me the others were merely suggestions that you are free to adjust to or leave it as is.
> 
> But let us double check: [@Daniel-Diaz](https://github.com/Daniel-Diaz) do you feel strongly about the wording? Otherwise I want to move to a vote.

Not strongly at all. It was just a suggestion. We can move on to the vote.

**Comment by hasufell** [comment]
Dear CLC members, let's vote on the proposal to add `{-# WARNING in "x-partial" ... #-}` to `Data.List.init` and `Data.List.last`, in similar fashion to head/tail... to warn users of their partial nature and suggest `NonEmpty` and `unsnoc` as alternatives.

The MR is here: https://gitlab.haskell.org/ghc/ghc/-/merge_requests/13564

----

@angerman @parsonsmatt @velveteer @TeofilC @Daniel-Diaz @ChickenProp @noughtmare @mpscholten @doyougnu

**Comment by TeofilC** [comment]
+1

**Comment by noughtmare** [comment]
+1

**Comment by hasufell** [comment]
+1

**Comment by treeowl** [comment]
I know I'm coming to this late, but there's a substantial alternative available for `init`: we could making it saturating, with `init [] = []`. I believe this would be better than adding a partiality warning. @Bodigrim , would it be possible to delay the vote slightly to allow for this possibility to be considered?

**Comment by hasufell** [comment]
@treeowl that's quite a big change.

In case you write such a proposal and it is accepted, we would of course be able to remove the partiality warning.

I suggest we keep going.

**Comment by doyougnu** [comment]
+1

**Comment by mpscholten** [comment]
👍

**Comment by velveteer** [comment]
+1

**Comment by Bodigrim** [comment]
@treeowl I'm no longer calling shots here, Julian is in charge now.

(I think we should rather add `dropEnd` / `takeEnd` / `splitAtEnd`, so that `init` can be replaced with `dropEnd 1`)

**Comment by ChickenProp** [comment]
+1

**Comment by hasufell** [comment]
8 votes out of 10 is enough to approve. Thanks all.

----

@mpilgrem thanks for your proposal!

Please now double check the post-approval point 6: https://github.com/haskell/core-libraries-committee/blob/main/PROPOSALS.md#the-how

----

@treeowl you're welcome to raise a new proposal for the change to `init`
